src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,306 @@
+        {
+            ISymbol symbol = context.Symbol;
+
+            // Don't report diagnostics on types that are marked as Preview
+            if (ProcessPreviewAttribute(symbol, requiresPreviewFeaturesSymbols))
+            {
+                return;
+            }
+
+            var reportDiagnostic = symbol switch
+            {
+                ITypeSymbol typeSymbol => TypeSymbolHasAttribute(typeSymbol, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType),
+                IPropertySymbol or IMethodSymbol => PropertyOrMethodSymbolHasAttribute(symbol, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType, runtimeFeatureType),
+                _ => false
+            };
+
+            if (reportDiagnostic)
+            {
+                if (requiresPreviewFeaturesSymbolsToUsageType.TryGetValue(symbol, out PreviewFeatureUsageType usageType))
+                {
+                    if (usageType == PreviewFeatureUsageType.StaticAbstract)
+                    {
+                        context.ReportDiagnostic(symbol.CreateDiagnostic(StaticAbstractIsPreviewFeatureRule));
+                    }
+                    else if (usageType == PreviewFeatureUsageType.EmptyInterface)
+                    {
+                        context.ReportDiagnostic(symbol.CreateDiagnostic(ImplementsEmptyPreviewInterfaceRule, symbol.Name));
+                    }
+                }
+                else
+                {
+                    context.ReportDiagnostic(symbol.CreateDiagnostic(GeneralPreviewFeatureAttributeRule, symbol.Name));
+                }
+            }
+        }
+
+        private static void BuildSymbolInformationFromOperations(OperationAnalysisContext context, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            if (OperationUsesPreviewFeatures(context, requiresPreviewFeaturesSymbols, out ISymbol? symbol))
+            {
+                IOperation operation = context.Operation;
+                if (operation is ICatchClauseOperation catchClauseOperation)
+                {
+                    operation = catchClauseOperation.ExceptionDeclarationOrExpression;
+                }
+
+                context.ReportDiagnostic(operation.CreateDiagnostic(GeneralPreviewFeatureAttributeRule, symbol.Name));
+            }
+        }
+
+        private static bool OperationUsesPreviewFeatures(OperationAnalysisContext context, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, [NotNullWhen(true)] out ISymbol? symbol)
+        {
+            IOperation operation = context.Operation;
+            symbol = GetOperationSymbol(operation);
+            if (operation is IInvocationOperation || operation is IPropertyReferenceOperation)
```

Comment:
Let me know if someone prefers the switch expression here. Since it's only 2 types here, I didn't bother with `switch`

Reply:
I wonder why it should apply only these 2, seems could apply for all operations

Reply:
Good point. When I wrote this first I wanted to narrow the scope, but I don't see why it shouldn't be all operations. Fixed it now by removing the `if` condition. This should now also work if say, a preview operator + was used inside another preview method.

Reply:
> Fixed it now by removing the if condition.

Do not see it is removed 😆

Reply:
Definitely removed now